File: sdks/python/src/echo_agent_sdk/permission_rule_matcher.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


class ToolPermission(str, Enum):
    READ = "read"
    WRITE = "write"
    NETWORK = "network"
    EXECUTE = "execute"
    SENSITIVE = "sensitive"

    def __str__(self) -> str:
        return self.value


class RuleMatcherKind(str, Enum):
    TOOL = "tool"
    PATTERN = "pattern"
    PERMISSION = "permission"
    ALL = "all"


@dataclass(frozen=True, slots=True)
class RuleMatcher:
    kind: RuleMatcherKind
    value: str | ToolPermission | None = None

    @classmethod
    def parse(cls, value: str) -> RuleMatcher:
        if not isinstance(value, str):
            raise TypeError("permission matcher must be text")
        if value == "*" or value == "all":
            return cls(RuleMatcherKind.ALL)
        if value.startswith("tool:"):
            name = value[5:]
            if not name:
                raise ValueError("tool permission matcher requires a name")
            return cls(RuleMatcherKind.TOOL, name)
        if value.startswith("pattern:"):
            pattern = value[8:]
            if not pattern:
                raise ValueError("pattern permission matcher cannot be empty")
            return cls(RuleMatcherKind.PATTERN, pattern)
        flag = (
            value[5:]
            if value.startswith("perm:")
            else value[11:]
            if value.startswith("permission:")
            else None
        )
        if flag is not None:
            try:
                return cls(RuleMatcherKind.PERMISSION, ToolPermission(flag))
            except ValueError as error:
                raise ValueError(f"unknown permission matcher: {flag}") from error
        raise ValueError(f"unsupported permission matcher: {value}")

    def display(self) -> str:
        if self.kind is RuleMatcherKind.TOOL:
            return f"tool:{self.value}"
        if self.kind is RuleMatcherKind.PATTERN:
            return f"pattern:{self.value}"
        if self.kind is RuleMatcherKind.PERMISSION:
            return f"permission:{self.value}"
        return "all"

    def matches_matcher_str(self, value: str) -> bool:
        return (
            self.kind in {RuleMatcherKind.TOOL, RuleMatcherKind.PATTERN}
            and self.value == value
        ) or (self.kind is RuleMatcherKind.ALL and value in {"*", "all"})

    def matches(
        self,
        tool_name: str,
        permissions: tuple[ToolPermission, ...] | list[ToolPermission],
    ) -> bool:
        if self.kind is RuleMatcherKind.TOOL:
            return tool_name == self.value
        if self.kind is RuleMatcherKind.PERMISSION:
            return self.value in permissions
        if self.kind is RuleMatcherKind.ALL:
            return True
        pattern = str(self.value)
        if pattern == "*":
            return True
        if tool_name == pattern:
            return True
        if _glob_matches(pattern, tool_name):
            return True
        if pattern.endswith("*)") and tool_name.startswith(pattern[:-2]):
            return True
        return (
            tool_name.startswith(pattern)
            and len(tool_name) > len(pattern)
            and tool_name[len(pattern)] == "("
        )


def _glob_matches(pattern: str, value: str) -> bool:
    pattern_index = 0
    value_index = 0
    star = -1
    mark = -1
    while value_index < len(value):
        if pattern_index < len(pattern) and (
            pattern[pattern_index] == "?"
            or pattern[pattern_index] == value[value_index]
        ):
            pattern_index += 1
            value_index += 1
        elif pattern_index < len(pattern) and pattern[pattern_index] == "*":
            star = pattern_index
            pattern_index += 1
            mark = value_index
        elif star != -1:
            pattern_index = star + 1
            mark += 1
            value_index = mark
        else:
            return False
    while pattern_index < len(pattern) and pattern[pattern_index] == "*":
        pattern_index += 1
    return pattern_index == len(pattern)

File: sdks/typescript/src/permission_rule_matcher.ts
/** Permission matcher values and pure matching helpers. */
import type { ToolPermission } from "./types.js";

export const RuleMatcherPermission = Object.freeze({
  Read: "read",
  Write: "write",
  Network: "network",
  Execute: "execute",
  Sensitive: "sensitive",
} as const);

export type RuleMatcher =
  | { readonly kind: "tool"; readonly name: string }
  | { readonly kind: "pattern"; readonly pattern: string }
  | { readonly kind: "permission"; readonly permission: ToolPermission }
  | { readonly kind: "all" };

export function ruleMatcherParse(value: string): RuleMatcher {
  if (typeof value !== "string") throw new TypeError("permission matcher must be text");
  if (value === "*" || value === "all") return Object.freeze({ kind: "all" });
  if (value.startsWith("tool:")) {
    const name = value.slice(5);
    if (!name) throw new Error("tool permission matcher requires a name");
    return Object.freeze({ kind: "tool", name });
  }
  if (value.startsWith("pattern:")) {
    const pattern = value.slice(8);
    if (!pattern) throw new Error("pattern permission matcher cannot be empty");
    return Object.freeze({ kind: "pattern", pattern });
  }
  const flag = value.startsWith("perm:") ? value.slice(5) : value.startsWith("permission:") ? value.slice(11) : undefined;
  if (flag !== undefined && Object.values(RuleMatcherPermission).includes(flag as ToolPermission)) {
    return Object.freeze({ kind: "permission", permission: flag as ToolPermission });
  }
  if (flag !== undefined) throw new Error(`unknown permission matcher: ${flag}`);
  throw new Error(`unsupported permission matcher: ${value}`);
}

export function ruleMatcherDisplay(matcher: RuleMatcher): string {
  switch (matcher.kind) {
    case "tool": return `tool:${matcher.name}`;
    case "pattern": return `pattern:${matcher.pattern}`;
    case "permission": return `permission:${matcher.permission}`;
    case "all": return "all";
  }
}

export function ruleMatcherMatchesMatcherStr(matcher: RuleMatcher, value: string): boolean {
  if (matcher.kind === "tool") return matcher.name === value;
  if (matcher.kind === "pattern") return matcher.pattern === value;
  return matcher.kind === "all" && (value === "*" || value === "all");
}

export function ruleMatcherMatches(matcher: RuleMatcher, toolName: string, permissions: readonly ToolPermission[]): boolean {
  switch (matcher.kind) {
    case "tool": return toolName === matcher.name;
    case "permission": return permissions.includes(matcher.permission);
    case "all": return true;
    case "pattern":
      if (toolName === matcher.pattern) return true;
      if (globMatches(matcher.pattern, toolName)) return true;
      if (matcher.pattern.endsWith("*)") && toolName.startsWith(matcher.pattern.slice(0, -2))) return true;
      return toolName.startsWith(matcher.pattern)
        && toolName.length > matcher.pattern.length
        && toolName[matcher.pattern.length] === "(";
  }
}

function globMatches(pattern: string, value: string): boolean {
  let p = 0;
  let v = 0;
  let star = -1;
  let mark = -1;
  while (v < value.length) {
    if (p < pattern.length && (pattern[p] === "?" || pattern[p] === value[v])) {
      p += 1;
      v += 1;
    } else if (p < pattern.length && pattern[p] === "*") {
      star = p++;
      mark = v;
    } else if (star !== -1) {
      p = star + 1;
      v = ++mark;
    } else {
      return false;
    }
  }
  while (p < pattern.length && pattern[p] === "*") p += 1;
  return p === pattern.length;
}
